tech-spec.md

```diff
@@ -0,0 +1,310 @@
+Examples:
+
+```bash
+./share.py --org=OHRI --outdir=~/ohri/ schema.csv data.xlsx
+./share.py schema.csv mysql+pymysql://scott:tiger@localhost/foo
+./share.py schema.csv mssql+pyodbc://user:pass@mydsn
+```
+
+## API
+
+### Public modules
+
+#### sharing
+
+- high level:
+    ```python
+    extract(data_source: str, schema_file: str, org: str=''
+            ) -> Dict[TableName, Dataframe]
+    ```
+    - data_source: a file path or database url (in SQLAlchemy format)
+    - schema_file: rule schema file path
+    - org: org to share with, or all orgs if empty
+    - returns a Pandas Dataframe per table, for further processing
+- low level:
+    - connect(data_source: str) -> Connection
+    - parse(schema_file: str) -> Dict[OrgName, Dict[TableName, Query]]
+    - extract(Connection, Query) -> Dataframe
+    - getCounts(Connection, Query) -> Dict[RuleId, int]
+    - getColumns(Connection, Query) -> Tuple[RuleId, List[ColumnName]]
+
+### Private modules
+
+#### cons
+
+Data source connection abstraction, including intermediate copy of spreadsheet
+files to in-memory DBs.
+
+- connect(data_source: str) -> Connection
+
+#### rules
+
+Loading of rule files.
+
+- load(schema_file: str) -> Dict[RuleId, Rule]
+
+#### trees
+
+Parsing of rules into abstract syntax trees.
+
+- parse(rules: Dict[RuleId, Rule] | List[Rule]) -> RuleTree
+
+#### queries
+
+(SQL) query generation from ASTs.
+
+- generate(rt: RuleTree) -> Dict[OrgName, Query]
+
+### Examples
+
+common definitions:
+```python
+import pandas as pd
+
+import sharing as s
+
+data_file = 'data.xlsx'
+rule_file = 'rules.csv'
+org = 'OHRI'
+```
+
+high-level one-shot function:
+
+```python
+results = s.extract(data_file, rules, org)
+for table, data in results.items():
+    data.to_csv(f'{org}-{table}.csv')
+```
+
+low-level sample code:
+
+```python
+def describeTableQuery(con, rules, table, query):
+    print(f'query table: {table}')
+
+    (selectRuleId, columns) = s.getColumns(con, query)
+    print(f'query columns (from rule {selectRuleId}):')
+    print(','.join(columns))
+
+    print('query counts per rule:')
+    ruleCounts = s.getCounts(con, query)
+    for ruleId, count in ruleCounts.items():
+        r = rules[ruleId]
+        ruleFilter = f'{r.key} {r.operator} {r.value}'
+        print(f'{ruleId} | {count} | {ruleFilter}')
+
+def extractFilteredData(con, table, query):
+    data: pd.Dataframe = s.extract(con, query)
+    data.to_csv(f'{org}-{table}.csv')
+
+con = s.connect(data_file)
+rules = s.load(rule_file)
+ruleTree = s.parse(rules)
+tableQueries = s.generate(ruleTree)[org]
+for table, query in tableQueries.items():
+    describeTableQuery(con, rules, table, query)
+    extractFilteredData(con, table, query)
+```
+
+## Rule schema parsing
+
+1. open schema file
+2. read lines
+3. parse each line into a rule obj
+4. add each rule obj to a dictionary with rule id as key
+
+## AST generation
```

Comment:
I think this makes more sense as a sub-section of the previous section (rule schema parsing).

Reply:
AST gen is completely separate from rule parsing as the spec is currently. The following three steps will be performed sequentially at the same level:

schema parsing -> ast gen -> sql gen

I will also add schema specific error reporting to the "rule schema parsing" section.

Reply:
done